### PR TITLE
Fixing decision procedure + better commented code

### PR DESCRIPTION
**Pull request by jurajsic**
I added more explanatory comments to decision procedure and tried to at least visually divide the different parts in it. Also a fix for the previous fix, empty sides should now propagate properly even for left side vars. There are also some more fixes (computing which new nodes are on cycle, or propagating that the change on the left in the same inclusion should add the same node to be processed again). I also removed inclusion graph from the decision procedure and moved all the inclusions keeping into the `SolvingState`, inclusion graph is now called only at the initialization.

### CODE REVIEW

**Comment by jurajsic** [comment]
I am gonna run some benchmarks to see if nothing is broken and then we can merge this.

**Comment by Adda0** [comment]
Let us know when the PR is finished and we can review the PR.

**Comment by jurajsic** [comment]
Ok, I run the benchmarks, it looks like it's working, you can check it now.

File: src/smt/theory_str_noodler/decision_procedure.cpp [hunk 1]
```diff
@@ -65,53 +102,72 @@ namespace smt::noodler {
...
+                tout << std::endl;
+            );
+
+            const auto &left_side_vars = inclusion_to_process.get_left_side();
+            const auto &right_side_vars = inclusion_to_process.get_right_side();
+
+            /********************************************************************************************************/
+            /****************************************** One side is empty *******************************************/
+            /********************************************************************************************************/
+            // As kinda opmtimalization step, we do "noodlification" for empty sides separately (i.e. sides that
```
Comment (Adda0) [inline src/smt/theory_str_noodler/decision_procedure.cpp:148]: Fix typo.
```suggestion
            // As kinda optimization step, we do "noodlification" for empty sides separately (i.e. sides that
```

**Review by Adda0 — COMMENTED** [review]
The changes so far seem good. Seeing that the benchmarks pass, I choose to believe these changes are sound.

I will approve the PR when it is finished (the fix for literals is pushed to the branch as well).

File: src/smt/theory_str_noodler/decision_procedure.h [hunk 2]
```diff
@@ -104,67 +104,151 @@ namespace smt::noodler {
...
-        std::unordered_map<BasicTerm, std::vector<BasicTerm>> substitution_map;
 
         SolvingState() = default;
         SolvingState(AutAssignment aut_ass,
-                        std::deque<std::shared_ptr<GraphNode>> nodes_to_process,
-                        std::shared_ptr<Graph> inclusion_graph,
-                        std::unordered_set<BasicTerm> length_sensitive_vars,
-                        std::unordered_map<BasicTerm, std::vector<BasicTerm>> substitution_map)
+                     std::deque<Predicate> inclusions_to_process,
+                     //std::shared_ptr<Graph> inclusion_graph,
```
[the remark this answers is not in the document]
Comment (jurajsic) [reply src/smt/theory_str_noodler/decision_procedure.h]: done

File: src/smt/theory_str_noodler/decision_procedure.h [hunk 3]
```diff
@@ -104,67 +104,151 @@ namespace smt::noodler {
...
-                          inclusion_graph(inclusion_graph),
-                          length_sensitive_vars(length_sensitive_vars),
-                          substitution_map(substitution_map) {}
-
-        // pushes node to the beginning of nodes_to_process but only if it is not in it yet
-        void push_front_unique(const std::shared_ptr<GraphNode> &node) {
-            if (std::find(nodes_to_process.begin(), nodes_to_process.end(), node) == nodes_to_process.end()) {
-                nodes_to_process.push_front(node);
+                          substitution_map(substitution_map),
+                          //inclusion_graph(inclusion_graph),
```
[the remark this answers is not in the document]
Comment (jurajsic) [reply src/smt/theory_str_noodler/decision_procedure.h]: done

File: src/smt/theory_str_noodler/inclusion_graph.cc [hunk 4]
```diff
@@ -239,10 +237,8 @@ Graph smt::noodler::Graph::create_inclusion_graph(Graph& simplified_splitting_gr
         for (auto& node: simplified_splitting_graph.get_nodes()) {
             if (simplified_splitting_graph.inverse_edges.count(node) == 0) {
                 inclusion_graph.nodes.insert(node);
+                STRACE("str", tout << "Added node " << node->get_predicate() << " to the graph without its brother." << std::endl;);
```
[the remark this answers is not in the document]
Comment (jurajsic) [reply src/smt/theory_str_noodler/inclusion_graph.cc]: Yes, I changed it

File: src/smt/theory_str_noodler/decision_procedure.cpp (hunk 1 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/decision_procedure.cpp]: fixed

**Comment by jurajsic** [comment]
We can merge, it looks like it works correctly

**Review by Adda0 — APPROVED** [review]

**Pull request merged**